source/slang/slang-type-layout.cpp

```diff
@@ -366,6 +366,25 @@ struct CPULayoutRulesImpl : DefaultLayoutRulesImpl
         }
     }
 
+    SimpleArrayLayoutInfo GetArrayLayout( SimpleLayoutInfo elementInfo, LayoutSize elementCount) override
+    {
+        if (elementCount.isInfinite())
+        {
+            // This is an unsized array, get information for element
+            auto info = Super::GetArrayLayout(elementInfo, LayoutSize(1));
+
+            // So it is actually a Array<T> on CPU which is a pointer and a size
+            info.size = sizeof(void*) * 2;
+            info.alignment = sizeof(void*);
+
+            return info;
+        }
```

Comment:
This works for right now, but in the long run I don't think we can get away with this simple logic.

The approach here means that *every* unsized array gets mapped to the pointer+count representation, no matter where it occurs. That means that, e.g., an unsized array as the last field of a `struct` type becomes an indirection, rather than a tail-allocated array.

I suspect that the longer-term fix here will be to leave the existing logic that computes the size of the array as infinite, and then special case such arrays (or unbounded-size parameters in general) when they occur at global scope in `slang-parameter-binding.cpp`.

To be clear: I'm 100% okay with this as a short-term solution to make global-scope unbounded arrays work for users of the CPU back-end. I just want to be clear that the longer-term solution might involve back out this change and doing something far more subtle.

Reply:
```That means that, e.g., an unsized array as the last field of a struct type becomes an indirection, rather than a tail-allocated array.```

True and that may is somewhat different to what happens on GPU. That I don't see any particular disadvantage to that tho - it is more general than just 'falling off the end', and allows bounds checking out of the box which may be very important for some uses. It also means it's use doesn't have the restrictions of use GPU might have. I can place unbounded arrays wherever I want for example. 

Now it could be argued there are reasons to want the falling off the end behavior - I don't know for better cache coherency. Or to be closer to how GPU works. I'm not sure that's particularly compelling tho. 

Perhaps the point is that there should be different syntax for this usage than the 'fall off the end style'.

Reply:
> True and that may is somewhat different to what happens on GPU. That I don't see any particular disadvantage to that tho - it is more general than just 'falling off the end', and allows bounds checking out of the box which may be very important for some uses.

We can make bounds checking work for the tail-allocated ("falling off the end") case as well, just by inserting an implicit field to account for the count, but I take your point that doing so is not as simple because somebody actually has to introduce that implicit/hidden field.

> It also means it's use doesn't have the restrictions of use GPU might have. I can place unbounded arrays wherever I want for example.

But (and this is the important point, maybe): as soon as you do that you aren't writing portable code any more, because you've written something that works on our CPU target and not on any of our current GPU targets.

> Now it could be argued there are reasons to want the falling off the end behavior - I don't know for better cache coherency. Or to be closer to how GPU works. I'm not sure that's particularly compelling tho.

The efficiency argument for me is less about caching and more about how tail allocation is a nice allocation policy (allocate one block instead of two), but of course you can do tail-allocation when you call `malloc` or `new`, but then still add the pointer+count in the header anyway.

The consistency with the GPU thing is more important in my mind, and is largely driven by the fact that "just add a level of indirection" is often cheap (in terms of cognitive load) on CPU, but expensive on current graphics APIs. If we always map this stuff to indirection on CPU, it is logical to do the same for GPU, but users may not be happy with the resulting cost, and will start wanting us to support the more efficient special cases, which amount to tail-allocation.

> Perhaps the point is that there should be different syntax for this usage than the 'fall off the end style'.

This gets back to the point where I initially advocated for using `ParameterBlock<...>` as the marker here, because a parameter block implies an indirection (on all targets that support indirection). That would mean there is a difference between `Foo[]` and `ParameterBlock<Foo[]>` because one is an unbounded-size type and the other is a bounded-size type.

Under that model, a user who *wants* the indirection can always ask for it, and it would be done using a portable feature (parameter blocks) so they would get a consistent result on CPU + GPU.

The only special-case behavior under the model I'm proposing would be that a global shader parameter with unbounded-size type would be implicitly "promoted" to a full parameter block (and thus get the indirection), unless it had specific annotations that tie our hands (e.g., a `register(...)` binding could force a different layout).

The problem is that there are two different reasonable policies here:

1) Allow unbounded-size types in tail position of `struct`s, where they will mean tail allocation ("contiguous")

2) Allow unbounded-size types in other positions, where they will mean indirection

Allowing just (1) is fine, and there is always `ParameterBlock` as an way to opt in to indirection.

Allowing just (2) is fine, but there would be no way to opt into tail-allocation (without inventing new syntax)

Allowing (1) + (2) starts to get slippery, and requires well-defined rules like "the behavior for fields is (1) and the behavior for globals is (2)." I don't think it would ever work well to say that all but the last field in a `struct` gets behavior (2), while the last field gets behavior (1).

The big caveat in all this is that GPU users end up having expectations about what is "obvious," and a lot of people would expect the following to mean tail-allocation in a single descriptor table/set:

```
struct MyStuff { Texture2D a; Texture2D b[]; }
ParameterBlock<MyStuff> gMyStuff;
```

If we tell GPU users that the above actually uses two descriptor tables/sets, then we need to be ready to justify *why* that is the case, and probably have a story for how to get the single-set behavior.

I initially assumed we'd want `gMyStuff` to map to two tables/sets and the implementation mostly tries to do that approach, but all my communication with users told me that the expectation was to do the other way, and the users changed my mind on this (at least when it comes to GPU behavior).